### ISSUE TITLE
Porting to Python3

### ISSUE THREAD

**Issue by mrv96** [tsmetana/python-libslp#3]
Hi, i'd like to use your python-libslp in a my project which use Python3.

I'm trying to port the code myself but i'm not a Python expert. At the moment your code (modified for Python3) compile but "import slp" in test.py fail because slp is not defined.

I guess that something in your installation procedure should be modified as done for the code, but i'm very inexpert.

Can you either make a porting of your code or help me in some other way?

Thank you very much.

**Comment by rchateauneu** [comment]
Hi

The module "slp" is a C module defined by the file slpmodule.c. Please see the line:

m = Py_InitModule("slp", slp_methods);

Porting a C module to Python 3 takes a couple of days of C programming. For example, the function:
PyMODINIT_FUNC initslp(void)

... must be replaced by something like:
PyMODINIT_FUNC PyInit_slp(void)

Hope it helps.

**Comment by rchateauneu** [comment]
By the way, you might be interested by this package which is in pure Python (No C language):

https://pypi.org/project/pyslp/
https://pythonhosted.org/pyslp/

**Comment by mrv96** [comment]
@rchateauneu thank you for your fast answer.

I have alrealdy done the modifies that you mentioned in the .c file and it compiles without error. The problem is when i try to include the module in the python test file.

For what concern pyslp I already know it, but the library is not 100% complete. I need to use a DA in my project and it seems that it isn't implemented in pyslp.

**Comment by tsmetana** [comment]
Hello. Thanks for your interest. The python-libslp was just a one shot proof-of-concept. Even though I think it worked fine, I haven't touched the code for quite some time as I don't have any use of it. I will accept any PRs. Even if you would want to fork and develop and maintain the code further, I really don't mind.

**Comment by rchateauneu** [comment]
Thanks Tomas for your proposal, this is very kind. Yes, it might be of interest.

mrv96: 
(1) For translating a Python2 module to Python3, more changes are needed if I remember well.
(2) Have you installed the newly created slp module ? You might set the environment variable PYTHONVERBOSE when attempting to import the module slp: Python will give a lot of details.

By the way, it is easy to ensure that a Python module written in C, is compatible both for Python2 and 3.

Also, it might be worth to create tests on Travis-ci and upload the latest version to https://pypi.org/project/pyslp/

**Comment by tsmetana** [comment]
So... I tried to compile and run the code out of curiosity. There were few changes required (this is really a very simple module) but it seems to be doing something on my system with Python 3.8.5.

@mrv96 feel free to test the code again.

**Comment by mrv96** [comment]
@tsmetana I've just forked your repo. I'm modifying the code. Yesterday i tried to rename slpmodule.c in slp.c and now the import works as well as the library that seems to work.

**Comment by tsmetana** [comment]
> @tsmetana I've just forked your repo. I'm modifying the code. Yesterday i tried to rename slpmodule.c in slp.c and now the import works as well as the library that seems to work.

This is odd: you should see undefined symbols even after renaming the DSO... But if it works... OK. :)

**Comment by mrv96** [comment]
> > @tsmetana I've just forked your repo. I'm modifying the code. Yesterday i tried to rename slpmodule.c in slp.c and now the import works as well as the library that seems to work.
> 
> This is odd: you should see undefined symbols even after renaming the DSO... But if it works... OK. :)

Sorry, i made a mistake. I renamed slpmodule.so in slp.so. The .c file was still slpmodule.c

**Comment by mrv96** [comment]
The last issue that i solved is with autoconf.

Since ./configure is not present i should run autoconf, but i receive this message:

configure.ac:5: error: possibly undefined macro: AM_INIT_AUTOMAKE
      If this token and others are legitimate, please use m4_pattern_allow.
      See the Autoconf documentation.

I solved this issue running autoreconf -fvi

**Comment by mrv96** [comment]
@tsmetana I've just finished to update my repo.

On my Ubuntu 18.04 my fork works, while yours not. I guess because in my configure.ac i specified python3 instead of python.

Is there a way to specify python version at compiling time without edit configure.ac?

**Comment by tsmetana** [comment]
Seems like Ubuntu is different: on Fedora the default python is version 3 so the configure script picks that one. I've added ifdefs to the code so it should compile on whatever version it finds. What is the exact error you get when trying to compile my code?

**Comment by mrv96** [comment]
During make command I receive normal not critical warnings:
![image](https://user-images.githubusercontent.com/13086678/93067959-26780980-f67c-11ea-961c-ed841b8304d3.png)

When i try to execute test.py I get this:
![image](https://user-images.githubusercontent.com/13086678/93068045-46a7c880-f67c-11ea-9ad7-2f0eaee6b618.png)

**Comment by rchateauneu** [comment]
Dear Mario

The error message indicates that it compiled initslp which implies Python-2
modules (init<module-name>).

This makes sense because the gcc command line indicates -I/.../python2.7

On the other hand, Python 3 modules must define PyInit_slp
(PyInit_<module-name>) and therefore Python3 complained that it could not
find it (That you can confirm with the command nm, for example, to see the
content of your shared library)..

Hope it helps.

**Comment by mrv96** [comment]
@rchateauneu your comment is very helpful.

I made a big mistake! Of course if I compile with python 2.7 i cannot execute test.py with python3.

I have just tried to execute test.py with python instead of python3 and it works well.

Although the version of python can be chosen changing configure.ac, now the problem is how to set the default python version on Ubuntu, but this isn't an issue related to this repo.

Thank you very much!
